File: app/src/main/java/io/nanodegree/andrea/popularmovies/presentation/detail/ReviewsAdapter.java
package io.nanodegree.andrea.popularmovies.presentation.detail;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import io.nanodegree.andrea.popularmovies.R;
import io.nanodegree.andrea.popularmovies.model.Review;

/**
 * Created by Andrea Loddo (andrea@evenly.io) on 11.05.2018
 * for Evenly GmbH
 * <p>
 * Copyright (c) 2018 Evenly GmbH,
 * all rights reserved
 */
public class ReviewsAdapter extends RecyclerView.Adapter<ReviewsAdapter.ReviewItemViewHolder> {

    private List<Review> reviewList;
    private Context context;

    ReviewsAdapter(@NonNull Context context) {
        this.context = context;
        reviewList = new ArrayList<>();
    }

    @NonNull
    @Override
    public ReviewItemViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        Context context = parent.getContext();
        LayoutInflater inflater = LayoutInflater.from(context);

        View view = inflater.inflate(R.layout.review_item, parent, false);

        return new ReviewItemViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ReviewItemViewHolder holder, int position) {
        holder.bind(reviewList.get(position));
    }

    @Override
    public int getItemCount() {
        return reviewList.size();
    }

    public void setData(List<Review> reviews) {
        this.reviewList.clear();
        this.reviewList.addAll(reviews);
    }

    class ReviewItemViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        TextView tv_review_title;
        TextView tv_review_content;
        Review review;

        ReviewItemViewHolder(View itemView) {
            super(itemView);

            tv_review_title = itemView.findViewById(R.id.review_title);
            tv_review_content = itemView.findViewById(R.id.review_content);
            itemView.setOnClickListener(this);
        }

        void bind(Review review) {
            this.review = review;
            tv_review_title.setText(review.author);
            tv_review_content.setText(review.content);
        }

        @Override
        public void onClick(View v) {

            Intent webIntent = new Intent(Intent.ACTION_VIEW, Uri.parse(review.url));
            context.startActivity(webIntent);
        }
    }
}

File: view/src/main/java/io/nanodegree/andrea/popularmovies/presentation/moviedetail/recyclerview/ReviewsAdapter.kt
package io.nanodegree.andrea.popularmovies.presentation.moviedetail.recyclerview

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.jetbrains.handson.mpp.mobile.ReviewContainer.Review
import io.nanodegree.andrea.popularmovies.R
import java.util.*

class ReviewsAdapter(private val context: Context) :
    RecyclerView.Adapter<ReviewsAdapter.ReviewItemViewHolder>() {

    private val reviewList: MutableList<Review> = ArrayList()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ReviewItemViewHolder {
        val context = parent.context
        val inflater = LayoutInflater.from(context)

        val view = inflater.inflate(R.layout.review_item, parent, false)

        return ReviewItemViewHolder(view)
    }

    override fun onBindViewHolder(holder: ReviewItemViewHolder, position: Int) {
        holder.bind(reviewList[position])
    }

    override fun getItemCount(): Int = reviewList.size

    fun setData(reviews: List<Review>) {
        this.reviewList.clear()
        this.reviewList.addAll(reviews)
    }

    inner class ReviewItemViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView),
        View.OnClickListener {

        private val reviewTitle: TextView = itemView.findViewById(R.id.review_title)
        private val reviewContent: TextView = itemView.findViewById(R.id.review_content)
        private var review: Review? = null

        init {
            itemView.setOnClickListener(this)
        }

        fun bind(review: Review) {
            this.review = review
            reviewTitle.text =
                String.format("%s %s", context.getString(R.string.detail_review_by), review.author)
            reviewContent.text = review.content
        }

        override fun onClick(v: View) {
            review?.let {
                val webIntent = Intent(Intent.ACTION_VIEW, Uri.parse(it.url))
                context.startActivity(webIntent)
            }
        }
    }
}
